circuit/src/gadgets/range.rs

```diff
@@ -0,0 +1,471 @@
+use crate::{
+	utils::{fe_to_le_bits, le_bits_to_u64},
+	RegionCtx,
+};
+use halo2::{
+	circuit::{AssignedCell, Layouter, Region, Value},
+	halo2curves::FieldExt,
+	plonk::{Constraints, Selector},
+	poly::Rotation,
+};
+
+use super::lt_eq_lookup::{MockChip, MockChipset, MockCommonConfig};
+
+/// Short length bits check chip using lookup table.
+///
+/// Checks if the target value is `S` bits long.
+#[derive(Debug, Clone)]
+pub struct LookupShortWordCheckChip<F: FieldExt, const K: usize, const S: usize> {
+	x: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt, const K: usize, const S: usize> LookupShortWordCheckChip<F, K, S> {
+	/// Construct new instance
+	pub fn new(x: AssignedCell<F, F>) -> Self {
+		Self { x }
+	}
+}
+
+impl<F: FieldExt, const K: usize, const S: usize> MockChip<F>
+	for LookupShortWordCheckChip<F, K, S>
+{
+	type Output = AssignedCell<F, F>;
+
+	fn configure(
+		mock_common: &MockCommonConfig, meta: &mut halo2::plonk::ConstraintSystem<F>,
+	) -> Selector {
+		assert!(0 < S && S < K, "Word bits should be less than target bits.");
+
+		let bitshift_selector = meta.complex_selector();
+
+		let word_column = mock_common.common.advice[0];
+
+		meta.lookup("Check K bits limit", |meta| {
+			let bitshift_selector = meta.query_selector(bitshift_selector);
+			let shifted_word = meta.query_advice(word_column, Rotation::cur());
+
+			vec![(bitshift_selector * shifted_word, mock_common.table)]
+		});
+
+		// For short lookups, check that the word has been shifted by the correct number
+		// of bits. https://p.z.cash/halo2-0.1:decompose-short-lookup
+		meta.create_gate("Short word S bits limit", |meta| {
+			let bitshift_selector = meta.query_selector(bitshift_selector);
+			let word = meta.query_advice(word_column, Rotation::prev());
+			let shifted_word = meta.query_advice(word_column, Rotation::cur());
+			let inv_two_pow_s = meta.query_advice(word_column, Rotation::next());
+
+			let two_pow_k = F::from(1 << K);
+
+			// shifted_word = word * 2^{K-s}
+			//              = word * 2^K * inv_two_pow_s
+			Constraints::with_selector(
+				bitshift_selector,
+				Some(word * two_pow_k * inv_two_pow_s - shifted_word),
+			)
+		});
+
+		bitshift_selector
+	}
+
+	fn synthesize(
+		self, mock_common: &MockCommonConfig, selector: &Selector, mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, halo2::plonk::Error> {
+		layouter.assign_region(
+			|| "short word check chip",
+			|region| {
+				let mut ctx = RegionCtx::new(region, 0);
+
+				// Assign original value
+				let assigned_x = ctx.copy_assign(mock_common.common.advice[0], self.x.clone())?;
+
+				ctx.next();
+
+				// Assign shifted value
+				let shifted_word = self.x.value().cloned() * Value::known(F::from(1 << (K - S)));
+				ctx.assign_advice(mock_common.common.advice[0], shifted_word)?;
+				ctx.enable(selector.clone())?;
+
+				ctx.next();
+
+				// Assign 2^{-S} from a fixed column.
+				let inv_two_pow_s = F::from(1 << S).invert().unwrap();
```

Comment:
The constant `1 << K` is defined as a contract expression: https://github.com/eigen-trust/protocol/pull/169/files#diff-11513d52971f0935b0b5cf0c36e0b15b6d9ef1595f05ed0915140a3cd5175dfbR58,  Why didnt you do the same with `inv_two_pow_s`

Reply:
This same constant is used in every row, so there is no need to store it inside the fixed column.
We can just hard code it into the constraint check.

Reply:
Done https://github.com/eigen-trust/protocol/pull/169/commits/a5db03110ec94caff1837ee9d9dda57ccfbac444

Please re-check the update.